src/shared/utils.ts

```diff
@@ -123,6 +123,9 @@ export function usingCustomScript(binPath: string, binArgs: string[], options?:
 			? options.customScriptReplacesNumArgs
 			: 1; // Default to removing one arg.
 		binArgs = binArgs.slice(numArgsToRemove);
+	} else if (fs.existsSync(flutterSnapBinPath) && binPath.endsWith("/flutter")) {
+		// If installed, use flutter from the snap instead of the SDK path.
+		binPath = flutterSnapBinPath;
```

Comment:
It feels like this change should be unnecessary - this code is specifically to handle user/repo-specified custom locations, and it should fall back to the detected SDK. If you remove this, what doesn't work correctly?

Reply:
If we remove this, the binPath will be set to `$HOME/snap/flutter/common/flutter/bin/flutter` and not `/snap/flutter/current/flutter.sh`.

Calling the former will result in build errors if any dev dependancies are not installed on the users system (for example, CMake or Ninja). This is the convenience provided by the snap: The latter is a script that ensures environment variables such as PATH, LDFLAGS, etc. are configured correctly to point into the snap package, such that the user need not manually do anything more than `snap install flutter` to get going with their Flutter development.

Reply:
> Calling the former will result in build errors if any dev dependancies are not installed on the user's system

Won't this have already been taken care of by the initialisation during the SDK searching?

Reply:
Build dependancies (CMake, Ninja, etc.) are installed together with the snap under `/snap/flutter/current/...`, however in order for flutter to know where those are, some environment variables need to be tweaked (PATH, LDFLAGS, etc.). `/snap/flutter/current/flutter.sh` does this. Calling `flutter` directly from the SDK directory does not.

Reply:
Oh, I thought this was just a script to do the initialisation. So the first time this script runs, it will initialise, and then subsequent runs it will forward on to the real Flutter binary? What's the output of the process on the first run?

I think it may be better to try and wrap it up in something like a `WorkspaceConfig` (similar to the Bazel one). The Bazel one overrides specific Flutter sub-commands scripts (`flutterRunScript` etc.) but there's no reason we couldn't also have a base `flutterScript` field that allows overriding the flutter binary rather than the sub-commands.

The entry point for the Bazel one is in `src/extension/sdk/utils.ts` and looks like this:

```ts
processWorkspaceType(findGitRoot, processKnownGitRepositories);
processWorkspaceType(findBazelWorkspaceRoot, (l, c, b) => processBazelWorkspace(l, c, b, config.previewBazelWorkspaceCustomScripts));
```

The first arguments (`findGitRoot`/`findBazelWorkspaceRoot`) detect whether a workspace is of a certain type (in this case it's not really workspace specific, but we can detect if the snap is installed) then the second argument (`progressKnownGitRepositories` etc.) will update the `WorkspaceConfig` (eg. in this case, set a new `flutterScript` field that would be used in the code above (this would make it generic instead of Snap-specific)).

I think that way, all snap-specific code is inside those two functions (find + process) rather than dotted across multiple places. If other OSes had similar things, this would also make it easier to implement them without extending all of these cases.

WDYT?

It may also be good to get an integration test around this (eg. on GitHub actions we use an Ubuntu box, install the snap, mock the prompt to automatically approve, and ensure we can load successfully), but we can do that once we've worked the implementation details out.

Thanks!

Reply:
Ha, it’s funny you mention doing this as a workspace config. That’s precisely how I originally implemented this and thought perhaps it was the wrong place :) I agree, it’s much cleaner that way, far less diff. Alright, I'll go in that direction again.

Reply:
Ah, doh :) I think it makes more sense wrapped up there - it was specifically to try and avoid leaking logic like this around the place, and instead allow a simpler way to override behaviour based on the environment.

The only possibly niggle, is that it might prevent us from suppressing the initialisation in the case where the user installed (but didn't initialize) the snap package, but then intends to use a VS-Code-configured path. It's probably a bit of an edge case though, so if the rest of the code is cleaner and isolated as a result, it's probably not a big problem.

Shout if anything isn't clear!

Reply:
Oh right. Yeah... it simply will not work without the explicit initialization logic as I have it now because the SDK path will not exist for the extension to detect. Chicken and egg situation.

Reply:
I was still thinking this would exist, but inside the `processSnapWorkspace`(?) function (so `processWorkspaceType` would likely need to be come async). So the equiv of the `find` function would return truthy if the Snap exists (whether it's initialised or not), then the equiv of the `process` function would initialise if required (being async so holding up the rest of the code while this happens), then write the paths into the config and continue.